src/Server/RestRouterHandlers/SearchHandler.cpp

```diff
@@ -0,0 +1,219 @@
+
+    LOG_INFO(log, "Execute query: {}", query);
+    ReadBufferFromString in{query};
+
+    /// Prepare output buffer
+    const auto out = getOutputBuffer(response);
+    try
+    {
+        /// FIXME: implement outbound solution to support query progress
+        executeQuery(in, *out, /* allow_into_outfile = */ false, query_context, {});
+    }
+    catch (Exception & e)
+    {
+        tryLogCurrentException(log);
+        /// Send the error message into already used (and possibly compressed) stream.
+        /// Note that the error message will possibly be sent after some data.
+        /// Also HTTP code 200 could have already been sent.
+        http_status = HTTPResponse::HTTP_INTERNAL_SERVER_ERROR;
+        response.setStatusAndReason(HTTPResponse::HTTPStatus(http_status));
+        bool data_sent = out->count() != out->offset();
+
+        /// If buffer has data, and that data wasn't sent yet, then no need to send that data
+        if (!data_sent)
+            out->position() = out->buffer().begin();
+
+        writeString(jsonErrorResponse(e.message(), e.code()), *out);
+        writeChar('\n', *out);
+        out->next();
+        out->finalize();
+    }
+
+    return "";
+}
+
+String SearchHandler::getQuery(const Poco::JSON::Object::Ptr & payload) const
+{
+    /// Setup query settings
+    const auto & mode = payload->has("mode") ? payload->get("mode").toString() : "standard";
+
+    /// FIXME: to support 'realtime' mode in future
+    /// FIXME: change the behavior of 'verbose' mode. Currently in frontier service, 'standard' and 'verbose' mode are the same
+    if (mode == "standard" || mode.empty() || mode == "verbose")
+        query_context->setDefaultFormat("JSONCompactEachRowWithNamesAndTypes");
+
+    const auto & start_time = payload->has("start_time") ? payload->get("start_time").toString() : "";
+    if (!start_time.empty())
+        query_context->setTimeParamStart(start_time);
+
+    const auto & end_time = payload->has("end_time") ? payload->get("end_time").toString() : "";
+    if (!end_time.empty())
+        query_context->setTimeParamEnd(end_time);
+
+    query_context->setSetting("unnest_subqueries", true);
+    query_context->setSetting("asterisk_include_materialized_columns", true);
+    query_context->setSetting("asterisk_include_alias_columns", true);
+
+    std::vector<String> query_parts;
+    query_parts.push_back(fmt::format("SELECT * FROM ({})", payload->get("query").toString()));
+    if (payload->has("offset") && payload->has("page_size"))
+    {
+        int offset = payload->getValue<int>("offset");
+        int page_size = payload->getValue<int>("page_size");
+        if (page_size > 0 && offset >= 0)
+            query_parts.push_back(fmt::format("LIMIT {}, {}", offset, page_size));
+    }
+    return boost::algorithm::join(query_parts, " ");
+}
+
+bool SearchHandler::validatePayload(const Poco::JSON::Object::Ptr & payload, String & error)
+{
+    error.clear();
```

Comment:
we don't need `clear`

Reply:
I think run clear is better to avoid the impact from last call with non empty error message.

Reply:
we are doing string assignment anyway ?